airflow/migrations/versions/142555e44c17_add_data_interval_start_end_to_dagrun.py

```diff
@@ -0,0 +1,155 @@
+
+    __tablename__ = "dag_run"
+
+    id = Column(Integer, primary_key=True)
+    dag_id = Column(String(ID_LEN))
+    execution_date = Column(UtcDateTime)
+    data_interval_start = Column(UtcDateTime)
+    data_interval_end = Column(UtcDateTime)
+    run_id = Column(String(ID_LEN))
+    run_type = Column(String(50))
+
+    dag = relationship(DagModel, primaryjoin=(foreign(DagModel.dag_id) == dag_id))
+
+
+class SerializedDagModel(Base):
+    """A partially frozen ``airflow.models.SerializedDagModel`` class."""
+
+    __tablename__ = "serialized_dag"
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    data = Column(JSONField(json=json))
+
+    @property
+    def dag(self):
+        """Copied from the original model class."""
+        SerializedDAG._load_operator_extra_links = self.load_op_links  # pylint: disable=protected-access
+        if isinstance(self.data, dict):
+            return SerializedDAG.from_dict(self.data)
+        return SerializedDAG.from_json(self.data)
+
+
+# These kinds of runs don't have a data interval and can be bulk-updated with
+# one SQL call, so we do them separately. Other "real" schedule intervals are
+# too complicated and need to be populated manually.
+NO_SCHEDULE_FILTER = or_(
+    DagModel.schedule_interval.in_([None, "@once"]),
+    DagRun.run_type == DagRunType.MANUAL,
+)
+
+
+def _populate_simple_dagrun_intervals(session):
+    """Handle DAG runs with simple schedule intervals."""
+    updates = {
+        DagRun.data_interval_start: DagRun.execution_date,
+        DagRun.data_interval_end: DagRun.execution_date,
+    }
+    # SQLite doesn't support UPDATE ... WHERE with multiple tables, and MySQL
+    # doesn't support subqeury in UPDATE, so...
+    if op.get_bind().dialect.name != 'sqlite':
+        dag_runs = session.query(DagRun).filter(DagRun.dag, NO_SCHEDULE_FILTER)
+    else:
+        dag_run_ids = session.query(DagRun.id).filter(NO_SCHEDULE_FILTER)
+        dag_runs = session.query(DagRun).filter(DagRun.id.in_(dag_run_ids.subquery()))
+    dag_runs.update(updates, synchronize_session=False)
+
+
+def _populate_complex_dagrun_intervals(session):
+    """Handle DAG runs with "real" schedule intervals."""
+    dag_runs_with_serialized = session.query(DagRun, SerializedDagModel).filter(
+        DagModel.dag_id == DagRun.dag_id,
+        SerializedDagModel.dag_id == DagRun.dag_id,
+        ~NO_SCHEDULE_FILTER,
+    )
+    for dag_run, serialized in dag_runs_with_serialized:
+        dag = serialized.dag
+        data_interval_start = dag_run.execution_date
+        dag_run.data_interval_start = data_interval_start
+        dag_run.data_interval_end = dag.following_schedule(data_interval_start)
+        session.merge(dag_run, load=False)
+
+
+def upgrade():
+    """Apply add data_interval_[start|end] to DagRun."""
+    # Create aolumns with NULL as default.
+    with op.batch_alter_table("dag_run") as batch_op:
+        batch_op.add_column(Column("data_interval_start", UtcDateTime))
```

Comment:
Is this the right type for mssql? Paging @aneesh-joseph

Reply:
Hmm CI for MSSQL is stuck during this migration so I likely am doing things wrong.

Reply:
MSSQL needs `mssql.DATETIME2(precision=6)`  for the column